arelle/plugin/validate/NL/rules/br_kvk.py

```diff
@@ -0,0 +1,251 @@
+    DISCLOSURE_SYSTEM_NT16,
+    DISCLOSURE_SYSTEM_NT17,
+)
+from ..PluginValidationDataExtension import PluginValidationDataExtension
+
+if TYPE_CHECKING:
+    from arelle import ModelXbrl
+    from arelle.ModelValue import QName
+
+_: TypeGetText
+
+
+def _getReportingPeriodDateValue(modelXbrl: ModelXbrl, qname: QName) -> date:
+    facts = modelXbrl.factsByQname.get(qname)
+    assert facts and len(facts) == 1, _('Exactly one fact must exist for reporting period concept: {0}').format(qname)
+    value = next(iter(facts)).value
+    return datetime.strptime(value, "%Y-%m-%d").date()
+
+
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+    disclosureSystems=[
+        DISCLOSURE_SYSTEM_NT16,
+        DISCLOSURE_SYSTEM_NT17,
+    ],
+)
+def rule_br_kvk_2_04(
+    pluginData: PluginValidationDataExtension,
+    val: ValidateXbrl,
+    *args: Any,
+    **kwargs: Any,
+) -> Iterable[Validation] | None:
+    """
+    BR-KVK-2.04: The period in the context MUST correspond to the period of the current or
+    previous financial reporting period, or one day before the start of the comparative financial
+    where the context must be instant.
+
+    Example:
+        Reporting Period Start Date: 2018-01-01
+        Reporting Period End Date: 2018-12-31
+
+    | Context | Instant    | Start Date | End Date   | Note                                                    |
+    | ------- | ---------- | ---------- | ---------- | ------------------------------------------------------- |
+    | 1       | 2018-12-31 |            |            | End value as context for the current reporting period   |
+    | 2       |            | 2018-12-31 | 2018-12-31 | Current reporting period                                |
+    | 3       | 2017-12-31 |            |            | End value as context for previous reporting period      |
+    | 4       |            | 2017-12-31 | 2017-12-31 | Previous reporting period                               |
+    | 5       | 2016-12-31 |            |            | Starting value as context for previous reporting period |
+
+    """
+    modelXbrl = val.modelXbrl
+
+    currentDuration = (
+        _getReportingPeriodDateValue(modelXbrl, FINANCIAL_REPORTING_PERIOD_CURRENT_START_DATE_QN),
+        _getReportingPeriodDateValue(modelXbrl, FINANCIAL_REPORTING_PERIOD_CURRENT_END_DATE_QN)
+    )
+    previousDuration = (
+        _getReportingPeriodDateValue(modelXbrl, FINANCIAL_REPORTING_PERIOD_PREVIOUS_START_DATE_QN),
+        _getReportingPeriodDateValue(modelXbrl, FINANCIAL_REPORTING_PERIOD_PREVIOUS_END_DATE_QN)
+    )
+    validDurations = (currentDuration, previousDuration)
+    validInstants = (
+        currentDuration[1],
+        (currentDuration[0] + timedelta(days=-1)),
+        (previousDuration[0] + timedelta(days=-1))
+    )
+
+    for contextId, context in modelXbrl.contexts.items():
+        if context.isInstantPeriod:
+            # instantDatetime getter adds a day, we need to subtract for comparison
```

Comment:
It seems like it would be clearer to add `instantDate` and `endDate` helpers that subtract a day if midnight or else return the day (per [4.7.2](https://www.xbrl.org/Specification/XBRL-2.1/REC-2003-12-31/XBRL-2.1-REC-2003-12-31+corrected-errata-2013-02-20.html#_4.7.2)).

Reply:
I believe the helpers just need to *not* pass in `addOneDay`, I've implemented them.

Reply:
Follow up [here](https://github.com/Arelle/Arelle/pull/876#discussion_r1338497169).